flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/source/split/reader/BigQuerySourceSplitReader.java

```diff
@@ -0,0 +1,335 @@
+                            .setReadStream(split.getStreamName())
+                            .setOffset(offsetToFetch(split))
+                            .build();
+
+            return client.readRows(readRequest);
+        } catch (Exception ex) {
+            throw new IOException(
+                    String.format(
+                            "[subtask #%d] Problems while opening the stream %s from BigQuery"
+                                    + " with connection info %s. Current split offset %d,"
+                                    + " reader offset %d.",
+                            readerContext.getIndexOfSubtask(),
+                            Optional.ofNullable(split.getStreamName()).orElse("NA"),
+                            readOptions.toString(),
+                            split.getOffset(),
+                            readSoFar),
+                    ex);
+        }
+    }
+
+    @Override
+    public RecordsWithSplitIds<GenericRecord> fetch() throws IOException {
+        if (closed) {
+            throw new IllegalStateException("Can't fetch records from a closed split reader.");
+        }
+
+        RecordsBySplits.Builder<GenericRecord> respBuilder = new RecordsBySplits.Builder<>();
+
+        // nothing to read has been assigned
+        if (assignedSplits.isEmpty()) {
+            return respBuilder.build();
+        }
+
+        // return when current read count is already over limit
+        if (readerContext.willItBeOverLimit(0)) {
+            LOG.info(
+                    "Completing reading because we are over limit (context reader count {}).",
+                    readerContext.currentReadCount());
+            respBuilder.addFinishedSplits(
+                    assignedSplits.stream()
+                            .map(split -> split.splitId())
+                            .collect(Collectors.toList()));
+            assignedSplits.clear();
+            return respBuilder.build();
+        }
+
+        BigQuerySourceSplit assignedSplit = assignedSplits.peek();
+        int maxRecordsPerSplitFetch = readOptions.getMaxRecordsPerSplitFetch();
+        int read = 0;
+        Long fetchStartTime = System.currentTimeMillis();
+        Boolean truncated = false;
+
+        try {
+            if (readStreamIterator == null) {
+                readStreamIterator = retrieveReadStream(assignedSplit).iterator();
+            }
+            Long itStartTime = System.currentTimeMillis();
+            while (readStreamIterator.hasNext()) {
+                ReadRowsResponse response = readStreamIterator.next();
+                if (!response.hasAvroRows()) {
+                    LOG.info(
+                            "[subtask #{}] The response contained no avro records for stream {}.",
+                            readerContext.getIndexOfSubtask(),
+                            assignedSplit.getStreamName());
+                }
+                if (avroSchema == null && response.hasAvroSchema()) {
```

Comment:
Simpler to read:
```
if (avroSchema == null) {
  if (response.hasAvroSchema()) {
     // this will happen only the first time we read from a particular stream
    avroSchema = new Schema.Parser().parse(response.getAvroSchema().getSchema());
  } else {
    throw new IllegalArgumentException(
                            "Avro schema not initialized and not available in the response.");
  }
}
```